### PR TITLE
Change copy read files parallel.

### LINKED ISSUE #4308: Improve the COPY from external location performance
**Summary**
If we COPY a s3 file and insert into a table, the progresses are:
S1. Read s3 file by blocks from s3 location
S2. Write blocks stream to table t1
S3. Commit

https://github.com/datafuselabs/databend/blob/16e06e414c4680f0d640abada631af89369be877/query/src/interpreters/interpreter_copy.rs#L83-L102

S1 and S2 is in the same thread, looks we can make them in parallel.

### PR DESCRIPTION
**Pull request by GrapeBaBa**
Signed-off-by: grapebaba <281165273@qq.com>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Change copy read files parallel.

## Changelog
- Performance Improvement

## Related Issues

Fixes #4308

### CODE REVIEW

File: query/src/interpreters/interpreter_copy.rs [hunk 1]
```diff
@@ -99,24 +100,22 @@ impl CopyInterpreter {
...
-        };
-
         let pipeline_builder = QueryPipelineBuilder::create(ctx.clone());
-        let mut pipeline = pipeline_builder.finalize(&from_plan)?;
+        let mut pipeline = NewPipeline::create();
+        for file in files {
+            let read_source_plan = self.plan.from.clone();
+            let read_source_plan = Self::rewrite_read_plan_file_name(read_source_plan, Some(file));
+            tracing::info!("copy_one_file_to_table: source plan:{:?}", read_source_plan);
+            let from_plan = common_planners::SelectPlan {
```
Comment (sundy-li) [inline query/src/interpreters/interpreter_copy.rs:114]: Oh, no need to build the SelectPlan.

Just use multiple

```
 let table = self.ctx.build_table_from_source_plan(plan)?;
        table.read2(self.ctx.clone(), plan, &mut self.pipeline)
```

File: query/src/pipelines/new/pipeline.rs [hunk 2]
```diff
@@ -43,6 +43,7 @@ use crate::pipelines::new::processors::ResizeProcessor;
 ///                                           +-----+                       +--->|Processor|
 ///                                                                              +---------+
 ///
+#[derive(Clone)]
```
Comment (zhang2014) [inline query/src/pipelines/new/pipeline.rs:46]: Why do we need this?

File: query/src/pipelines/new/pipeline_builder.rs [hunk 3]
```diff
@@ -52,9 +52,10 @@ use crate::sessions::QueryContext;
 /// # let builder = QueryPipelineBuilder::create(ctx);
 /// # let pipeline = builder.finalize(plan)?;
 /// ```
+#[derive(Clone)]
```
Comment (zhang2014) [inline query/src/pipelines/new/pipeline_builder.rs:55]: Why do we need this?

File: query/src/pipelines/new/pipeline_builder.rs [hunk 4]
```diff
@@ -52,9 +52,10 @@ use crate::sessions::QueryContext;
...
 pub struct QueryPipelineBuilder {
     ctx: Arc<QueryContext>,
-    pipeline: NewPipeline,
+    pipeline: Box<NewPipeline>,
```
Comment (zhang2014) [inline query/src/pipelines/new/pipeline_builder.rs:58]: Why do we need this?

File: query/src/interpreters/interpreter_copy.rs [hunk 5]
```diff
@@ -99,24 +100,22 @@ impl CopyInterpreter {
...
+                input: Arc::new(PlanNode::ReadSource(read_source_plan)),
+            };
+            pipeline = pipeline_builder.clone().finalize(&from_plan)?;
```
Comment (zhang2014) [inline query/src/interpreters/interpreter_copy.rs:117]: For each file, we will build a new pipeline, but only take the last pipeline?

File: query/src/pipelines/new/pipeline.rs (hunk 2 above)
Comment (GrapeBaBa) [reply query/src/pipelines/new/pipeline.rs]: Wrong code

File: query/src/pipelines/new/pipeline_builder.rs (hunk 4 above)
Comment (GrapeBaBa) [reply query/src/pipelines/new/pipeline_builder.rs]: Wrong code

File: query/src/interpreters/interpreter_copy.rs (hunk 5 above)
Comment (GrapeBaBa) [reply query/src/interpreters/interpreter_copy.rs]: Wrong code

**Comment by GrapeBaBa** [comment]
[nohup.txt](https://github.com/datafuselabs/databend/files/8458940/nohup.txt)
I got error when executing stateful test in local, the log fie is pasted.

File: query/src/interpreters/interpreter_copy.rs [hunk 6]
```diff
@@ -99,24 +98,22 @@ impl CopyInterpreter {
...
-        tracing::info!("copy_one_file_to_table: source plan:{:?}", read_source_plan);
-
-        let from_plan = common_planners::SelectPlan {
-            input: Arc::new(PlanNode::ReadSource(read_source_plan)),
-        };
-
-        let pipeline_builder = QueryPipelineBuilder::create(ctx.clone());
-        let mut pipeline = pipeline_builder.finalize(&from_plan)?;
+        let mut pipeline = NewPipeline::create();
+        for file in files {
```
Comment (zhang2014) [inline query/src/interpreters/interpreter_copy.rs:107]: We not should create a new pipe for each file. Maybe we should support multiple file reading in S3StageTable, and then push multiple files down to S3StageTable? reference: MemoryTable::read2

Comment (zhang2014) [reply query/src/interpreters/interpreter_copy.rs:107]: > [nohup.txt](https://github.com/datafuselabs/databend/files/8458940/nohup.txt) I got error when executing stateful test in local, the log fie is pasted.

Don't create a pipe for each file, which causes us to build a wrong pipeline.

Comment (GrapeBaBa) [reply query/src/interpreters/interpreter_copy.rs:107]: ok, let me check S3StageTable and MemoryTable impl.

File: query/src/storages/s3/s3_stage_table.rs [hunk 7]
```diff
@@ -88,22 +90,31 @@ impl Table for S3StageTable {
         _plan: &ReadDataSourcePlan,
         pipeline: &mut NewPipeline,
     ) -> Result<()> {
+        let settings = ctx.get_settings();
+        let mut builder = SourcePipeBuilder::create();
         let table_info = &self.table_info;
         let schema = table_info.schema.clone();
+        let mut files_deque = VecDeque::with_capacity(table_info.files.len());
+        for f in &table_info.files {
```
Comment (sundy-li) [inline query/src/storages/s3/s3_stage_table.rs:98]: 👍

**Review by sundy-li — APPROVED** [review]
LGTM

**Pull request merged**